File: downloads/code/RollingDice.java
import java.util.ArrayList;
import java.util.Iterator;
 
public class DiceDistribution {
    static Double[][] dist;
    static int DiceNum = 3;
    // dist[Dice Number][Number on dice]
 
    public static void main(String[] args) {
        buildDistRandomly();
 
        ArrayList<String> candidate = permutationWtihDupe();
        for (String can : candidate) {
            printItemProb(can);
        }
    }
 
    public static void printItemProb(String item){
        Double prob = 1.0;
        System.out.print(item + " : ");
        for ( int i = 0 ; i < item.length() ; i++){
            int num = Character.getNumericValue(item.charAt(i));
            prob *= dist[i][num-1];
        }
        System.out.println(prob);
    }
 
    public static void buildDistRandomly(){
        dist = new Double[DiceNum][6];
        for (int i = 0 ; i < DiceNum ; i++){
            dist[i] = buildDiceRandomly();
        }
    }
 
    public static Double[] buildDiceRandomly(){
        Double[] dist = new Double[6];
        Double sum = 0.0;
        for (int i = 0; i < 6 ; i++){
            Double rand = Math.random();
            sum += rand;
            dist[i] = rand;
        }
        for (int i = 0 ; i < 6 ; i++){
            dist[i] = dist[i] / sum;
        }
        return dist;
    }
 
    static ArrayList<String> permutationWtihDupe(){
        ArrayList<String> candidate = new ArrayList<String>();
        // init first candidate
        for (int i = 0; i < 6 ; i++){
            candidate.add(Integer.toString(i+1));
        }
        int curLen = 1;
 
        while (curLen < DiceNum) {
            ArrayList<String> newCandidate = new ArrayList<String>();
            for (Iterator<String> iter = candidate.iterator(); iter.hasNext();){
                String item = iter.next();
                for (int i = 0; i < 6 ; i++){
                    newCandidate.add( item + Integer.toString(i+1));
                }
            }
            candidate = newCandidate;
            curLen++;
        }
        return candidate;
    }
}

File: downloads/code/RollingDice.scala
import scala.util.Random
 
object rollingDice extends App {
  val diceNum = 3   //  number of your dices
  val dices = (1 to diceNum).map(_ => unbalancedDice()).toList
  printOutcome(dices, 0, 1)
   
  def unbalancedDice () : List[Double] = { 
    var dices = List[Double]()
    var sum = 0.0
    for ( i <- 1 to 6) {
      val ran = Random.nextDouble()
      sum += ran
      dices = ran :: dices
    }
    dices.map( x => x / sum)
  }
 
  def printOutcome(dices : List[List[Double]], diceIndex: Int, prob: Double): Unit = {
    for (i <- 0 to 5) {
      if (diceIndex == diceNum - 1)
        println(prob * dices(diceIndex)(i))
      else
        printOutcome(dices, diceIndex + 1, prob * dices(diceIndex)(i))
    }
  }
}
